File: twitter-anti-muzzle/typescript/antiMuzzle.ts
/**
 * Twitter Anti-Muzzle Utilities
 * Prevent your X/Twitter bot from getting muzzled (write access revoked)
 *
 * Usage:
 *   import { AntiMuzzle, ContentVariation } from './antiMuzzle';
 *
 *   const antiMuzzle = new AntiMuzzle();
 *
 *   if (antiMuzzle.canReply(userId, isBot)) {
 *     await antiMuzzle.addHumanDelay();
 *     const result = await api.postTweet(text);
 *     antiMuzzle.recordReply(userId);
 *   }
 */

interface AntiMuzzleConfig {
  maxRepliesHuman?: number;
  maxRepliesBot?: number;
  replyWindowHours?: number;
  vipAccounts?: Set<string>;
}

interface TokenItem {
  symbol: string;
  price: string;
  change: string;
}

/**
 * Anti-muzzling utilities for X/Twitter bots
 */
export class AntiMuzzle {
  private maxRepliesHuman: number;
  private maxRepliesBot: number;
  private replyWindowMs: number;
  private vipAccounts: Set<string>;
  private userReplies: Map<string, Date[]>;

  constructor(config: AntiMuzzleConfig = {}) {
    this.maxRepliesHuman = config.maxRepliesHuman ?? 5;
    this.maxRepliesBot = config.maxRepliesBot ?? 1;
    this.replyWindowMs = (config.replyWindowHours ?? 1) * 60 * 60 * 1000;
    this.vipAccounts = config.vipAccounts ?? new Set();
    this.userReplies = new Map();
  }

  /**
   * Check if we can reply to this user without hitting rate limits
   */
  canReply(userId: string, isBot: boolean = false): boolean {
    // VIP bypass
    if (this.vipAccounts.has(userId)) {
      return true;
    }

    const now = new Date();
    const cutoff = new Date(now.getTime() - this.replyWindowMs);

    // Clean old entries and count recent replies
    const replies = this.userReplies.get(userId) ?? [];
    const recentReplies = replies.filter((date) => date > cutoff);
    this.userReplies.set(userId, recentReplies);

    const recentCount = recentReplies.length;
    const maxAllowed = isBot ? this.maxRepliesBot : this.maxRepliesHuman;
    const canReply = recentCount < maxAllowed;

    if (!canReply) {
      console.log(
        `Rate limit: ${userId} (${isBot ? "bot" : "human"}) - ${recentCount}/${maxAllowed}`
      );
    }

    return canReply;
  }

  /**
   * Record that we replied to this user
   */
  recordReply(userId: string): void {
    const replies = this.userReplies.get(userId) ?? [];
    replies.push(new Date());
    this.userReplies.set(userId, replies);
  }

  /**
   * Add human-like delay before posting (simulates reading + thinking + typing)
   */
  async addHumanDelay(
    minSeconds: number = 3.0,
    maxSeconds: number = 8.0
  ): Promise<void> {
    const delay = Math.random() * (maxSeconds - minSeconds) + minSeconds;
    console.log(`Human-like delay: ${delay.toFixed(2)}s`);
    await this.sleep(delay * 1000);
  }

  /**
   * Add delay between chained posts (multiple images/tweets)
   */
  async addChainDelay(
    minSeconds: number = 8.0,
    maxSeconds: number = 15.0
  ): Promise<void> {
    const delay = Math.random() * (maxSeconds - minSeconds) + minSeconds;
    console.log(`Chain post delay: ${delay.toFixed(2)}s`);
    await this.sleep(delay * 1000);
  }

  /**
   * Add user to VIP list (bypass rate limits)
   */
  addVip(userId: string): void {
    this.vipAccounts.add(userId);
    console.log(`Added VIP: ${userId}`);
  }

  /**
   * Remove user from VIP list
   */
  removeVip(userId: string): void {
    this.vipAccounts.delete(userId);
    console.log(`Removed VIP: ${userId}`);
  }

  private sleep(ms: number): Promise<void> {
    return new Promise((resolve) => setTimeout(resolve, ms));
  }
}

/**
 * Rate-limit @mentions in outgoing reply text to avoid muzzling.
 *
 * Each @handle can only be mentioned once per cooldown period. When on
 * cooldown, the @handle is replaced with the account's display name
 * (or username without the @ if display name is unknown).
 *
 * Battle-tested on @solvrbot — this is the #1 defense against muzzling.
 */
export class MentionRateLimiter {
  private cooldownMs: number;
  private botHandle: string;
  private cooldowns: Map<string, number>; // handle_lower -> last_mentioned timestamp
  private displayNames: Map<string, string>; // handle_lower -> display name

  constructor(config: { cooldownSeconds?: number; botHandle?: string } = {}) {
    this.cooldownMs = (config.cooldownSeconds ?? 3600) * 1000;
    this.botHandle = (config.botHandle ?? "@mybot").toLowerCase();
    this.cooldowns = new Map();
    this.displayNames = new Map();
  }

  /**
   * Cache a display name for a handle (populate from X API responses)
   */
  cacheDisplayName(username: string, displayName: string): void {
    this.displayNames.set(`@${username.toLowerCase()}`, displayName);
  }

  /**
   * Bulk-cache display names from X API user includes
   */
  cacheFromApiUsers(users: Array<{ username: string; name?: string }>): void {
    for (const user of users) {
      if (user.username) {
        this.displayNames.set(
          `@${user.username.toLowerCase()}`,
          user.name ?? user.username
        );
      }
    }
  }

  /**
   * Rate-limit @mentions in outgoing text.
   * Returns text with cooldown-limited @mentions replaced by display names.
   */
  process(text: string): string {
    if (!text) return text;

    const now = Date.now();

    return text.replace(/@([A-Za-z0-9_]{1,15})\b/g, (handle) => {
      const handleLower = handle.toLowerCase();

      // Never tag ourselves
      if (handleLower === this.botHandle) {
        return this.displayNames.get(handleLower) ?? handle.slice(1);
      }

      const lastUsed = this.cooldowns.get(handleLower) ?? 0;
      if (now - lastUsed < this.cooldownMs) {
        // On cooldown — use display name or username without @
        return this.displayNames.get(handleLower) ?? handle.slice(1);
      } else {
        // Allow mention, start cooldown
        this.cooldowns.set(handleLower, now);
        return handle;
      }
    });
  }
}

/**
 * Utilities for varying response content to avoid repetitive patterns
 */
export class ContentVariation {
  /**
   * Add varied introduction to responses
   */
  static varyIntro(baseText: string): string {
    const intros = [
      "Here's what I found:",
      "Check these out:",
      "Here you go:",
      "Found these:",
      "Take a look:",
      "", // Sometimes no intro
    ];

    const intro = intros[Math.floor(Math.random() * intros.length)];
    return intro ? `${intro}\n\n${baseText}` : baseText;
  }

  /**
   * Format a list with varied structure
   */
  static varyListFormat(items: TokenItem[]): string {
    const formats = [
      (i: TokenItem) => `• ${i.symbol} - $${i.price} (${i.change})`,
      (i: TokenItem) => `${i.symbol}: $${i.price} | ${i.change}`,
      (i: TokenItem) => `$${i.symbol} ${i.price} ${i.change}`,
      (i: TokenItem) => `${i.symbol} ${i.price} · ${i.change}`,
    ];

    const formatter = formats[Math.floor(Math.random() * formats.length)];
    return items.map(formatter).join("\n");
  }

  /**
   * Return varied emoji selection
   */
  static varyEmojis(baseEmojis: string[]): string[] {
    const count = Math.floor(Math.random() * Math.min(3, baseEmojis.length + 1));

    if (count === 0) return [];

    // Fisher-Yates shuffle and take first N
    const shuffled = [...baseEmojis];
    for (let i = shuffled.length - 1; i > 0; i--) {
      const j = Math.floor(Math.random() * (i + 1));
      [shuffled[i], shuffled[j]] = [shuffled[j], shuffled[i]];
    }

    return shuffled.slice(0, count);
  }
}

/**
 * Decorator to monitor 403 Forbidden errors (early muzzling warning)
 */
export function monitor403Errors<T extends (...args: any[]) => Promise<any>>(
  target: any,
  propertyName: string,
  descriptor: TypedPropertyDescriptor<T>
) {
  const originalMethod = descriptor.value!;

  descriptor.value = async function (...args: any[]) {
    try {
      return await originalMethod.apply(this, args);
    } catch (error: any) {
      // Check for 403 Forbidden
      if (
        error?.message?.includes("403") ||
        error?.message?.includes("Forbidden")
      ) {
        console.error(`403 Forbidden detected: ${error.message}`);

        // Check if potentially muzzled
        if (
          error?.message?.toLowerCase().includes("muzzled") ||
          error?.message?.toLowerCase().includes("write")
        ) {
          console.error(
            "⚠️ MUZZLING DETECTED! Check X developer portal immediately."
          );
        }
      }

      throw error;
    }
  } as T;

  return descriptor;
}

// Example usage
if (require.main === module) {
  (async () => {
    // Initialize
    const antiMuzzle = new AntiMuzzle({
      maxRepliesHuman: 5,
      maxRepliesBot: 1,
      vipAccounts: new Set(["123456789"]), // VIP user IDs
    });

    const userId = "987654321";
    const isBot = false;

    // Check if we can reply
    if (antiMuzzle.canReply(userId, isBot)) {
      console.log(`✅ Can reply to ${userId}`);

      // Add human-like delay
      console.log("Adding human-like delay...");
      await antiMuzzle.addHumanDelay();

      // Simulate posting
      console.log("Posting tweet...");
      // await api.postTweet(text);

      // Record the reply
      antiMuzzle.recordReply(userId);
      console.log("Reply recorded");
    } else {
      console.log(`❌ Rate limited for ${userId}`);
    }

    // Content variation example
    const items: TokenItem[] = [
      { symbol: "BTC", price: "45000", change: "+5%" },
      { symbol: "ETH", price: "2500", change: "+3%" },
    ];

    const variedContent = ContentVariation.varyListFormat(items);
    console.log(`\nVaried content:\n${variedContent}`);

    // Mention rate limiter example
    const mentionLimiter = new MentionRateLimiter({
      cooldownSeconds: 3600,
      botHandle: "@mybot",
    });

    // Cache display names (from X API)
    mentionLimiter.cacheDisplayName("bankrbot", "Bankr");
    mentionLimiter.cacheDisplayName("virtuals_io", "Virtuals Protocol");

    // First mention — allowed
    const text1 = mentionLimiter.process("Check out @bankrbot for trading!");
    console.log(`\nFirst mention: ${text1}`);
    // -> "Check out @bankrbot for trading!"

    // Second mention within cooldown — replaced with display name
    const text2 = mentionLimiter.process("Also try @bankrbot for swaps!");
    console.log(`Second mention: ${text2}`);
    // -> "Also try Bankr for swaps!"
  })();
}

File: twitter-anti-muzzle/python/anti_muzzle.py
"""
Twitter Anti-Muzzle Utilities
Prevent your X/Twitter bot from getting muzzled (write access revoked)

Usage:
    from anti_muzzle import AntiMuzzle

    # Initialize
    anti_muzzle = AntiMuzzle()

    # Before posting
    if anti_muzzle.can_reply(user_id, is_bot=False):
        await anti_muzzle.add_human_delay()
        result = api.post_tweet(text)
        anti_muzzle.record_reply(user_id)
"""

import asyncio
import random
import time
import logging
from collections import defaultdict
from datetime import datetime, timedelta
from typing import Optional

logger = logging.getLogger(__name__)


class AntiMuzzle:
    """Anti-muzzling utilities for X/Twitter bots."""

    def __init__(
        self,
        max_replies_human: int = 5,
        max_replies_bot: int = 1,
        reply_window_hours: int = 1,
        vip_accounts: Optional[set[str]] = None,
    ):
        """
        Initialize anti-muzzle protection.

        Args:
            max_replies_human: Max replies per human user per window (default: 5)
            max_replies_bot: Max replies per bot account per window (default: 1)
            reply_window_hours: Time window for rate limiting (default: 1 hour)
            vip_accounts: Set of user IDs that bypass rate limits
        """
        self.max_replies_human = max_replies_human
        self.max_replies_bot = max_replies_bot
        self.reply_window = timedelta(hours=reply_window_hours)
        self.vip_accounts = vip_accounts or set()

        # Track recent replies per user
        self.user_replies: dict[str, list[datetime]] = defaultdict(list)

    def can_reply(self, user_id: str, is_bot: bool = False) -> bool:
        """
        Check if we can reply to this user without hitting rate limits.

        Args:
            user_id: Twitter user ID
            is_bot: Whether the user is a bot/AI agent

        Returns:
            True if we can reply, False if rate-limited
        """
        # VIP bypass
        if user_id in self.vip_accounts:
            return True

        now = datetime.utcnow()
        cutoff = now - self.reply_window

        # Clean old entries and count recent replies
        self.user_replies[user_id] = [
            t for t in self.user_replies[user_id] if t > cutoff
        ]
        recent_count = len(self.user_replies[user_id])

        # Check limit
        max_allowed = self.max_replies_bot if is_bot else self.max_replies_human
        can_reply = recent_count < max_allowed

        if not can_reply:
            logger.info(
                f"Rate limit: {user_id} ({'bot' if is_bot else 'human'}) "
                f"- {recent_count}/{max_allowed} in last {int(self.reply_window.total_seconds()) // 3600}h"
            )

        return can_reply

    def record_reply(self, user_id: str):
        """Record that we replied to this user."""
        self.user_replies[user_id].append(datetime.utcnow())

    async def add_human_delay(
        self, min_seconds: float = 3.0, max_seconds: float = 8.0
    ):
        """
        Add human-like delay before posting.

        Simulates "reading + thinking + typing" time.

        Args:
            min_seconds: Minimum delay in seconds
            max_seconds: Maximum delay in seconds
        """
        delay = random.uniform(min_seconds, max_seconds)
        logger.debug(f"Human-like delay: {delay:.2f}s")
        await asyncio.sleep(delay)

    async def add_chain_delay(
        self, min_seconds: float = 8.0, max_seconds: float = 15.0
    ):
        """
        Add delay between chained posts (multiple images/tweets).

        Longer delays to avoid rapid-fire posting patterns.

        Args:
            min_seconds: Minimum delay in seconds
            max_seconds: Maximum delay in seconds
        """
        delay = random.uniform(min_seconds, max_seconds)
        logger.debug(f"Chain post delay: {delay:.2f}s")
        await asyncio.sleep(delay)

    def add_vip(self, user_id: str):
        """Add user to VIP list (bypass rate limits)."""
        self.vip_accounts.add(user_id)
        logger.info(f"Added VIP: {user_id}")

    def remove_vip(self, user_id: str):
        """Remove user from VIP list."""
        self.vip_accounts.discard(user_id)
        logger.info(f"Removed VIP: {user_id}")


class MentionRateLimiter:
    """Rate-limit @mentions in outgoing reply text to avoid muzzling.

    Each @handle can only be mentioned once per cooldown period. When on
    cooldown, the @handle is replaced with the account's display name
    (or username without the @ if display name is unknown).

    Battle-tested on @solvrbot — this is the #1 defense against muzzling.
    """

    def __init__(
        self,
        cooldown_seconds: int = 3600,
        bot_handle: str = "@mybot",
    ):
        """
        Args:
            cooldown_seconds: How long before a handle can be @mentioned again (default: 1 hour)
            bot_handle: Your bot's handle (always stripped from replies)
        """
        self.cooldown_seconds = cooldown_seconds
        self.bot_handle = bot_handle.lower()
        self._cooldowns: dict[str, float] = {}  # handle_lower -> last_mentioned
        self._display_names: dict[str, str] = {}  # handle_lower -> display name

    def cache_display_name(self, username: str, display_name: str):
        """Cache a display name for a handle (populate from X API responses).

        Call this when processing mentions — X API includes user data with
        user_fields=["username", "name"].
        """
        self._display_names[f"@{username.lower()}"] = display_name

    def cache_from_api_users(self, users: list[dict]):
        """Bulk-cache display names from X API user includes.

        Args:
            users: List of user dicts with 'username' and 'name' keys
        """
        for user in users:
            username = user.get("username", "")
            name = user.get("name", username)
            if username:
                self._display_names[f"@{username.lower()}"] = name

    def process(self, text: str) -> str:
        """Rate-limit @mentions in outgoing text.

        Args:
            text: The reply text to process

        Returns:
            Text with rate-limited @mentions replaced by display names
        """
        if not text:
            return text

        import re
        now = time.time()

        def _replace(match: re.Match) -> str:
            handle = match.group(0)  # e.g. "@bankrbot"
            handle_lower = handle.lower()

            # Never tag ourselves
            if handle_lower == self.bot_handle:
                username = handle[1:]
                return self._display_names.get(handle_lower, username)

            last_used = self._cooldowns.get(handle_lower, 0)
            if now - last_used < self.cooldown_seconds:
                # On cooldown — use display name or just username without @
                username = handle[1:]
                replacement = self._display_names.get(handle_lower, username)
                logger.debug(f"Mention rate-limited: {handle} -> {replacement}")
                return replacement
            else:
                # Allow this mention and start cooldown
                self._cooldowns[handle_lower] = now
                logger.debug(f"Mention allowed: {handle} (cooldown started)")
                return handle

        return re.sub(r'@([A-Za-z0-9_]{1,15})\b', _replace, text)


class ContentVariation:
    """Utilities for varying response content to avoid repetitive patterns."""

    @staticmethod
    def vary_intro(base_text: str) -> str:
        """Add varied introduction to responses."""
        intros = [
            "Here's what I found:",
            "Check these out:",
            "Here you go:",
            "Found these:",
            "Take a look:",
            "",  # Sometimes no intro
        ]
        intro = random.choice(intros)
        return f"{intro}\n\n{base_text}" if intro else base_text

    @staticmethod
    def vary_list_format(items: list[dict]) -> str:
        """
        Format a list with varied structure.

        Args:
            items: List of dicts with 'symbol', 'price', 'change' keys

        Returns:
            Formatted string with randomized structure
        """
        formats = [
            lambda i: f"• {i['symbol']} - ${i['price']} ({i['change']})",
            lambda i: f"{i['symbol']}: ${i['price']} | {i['change']}",
            lambda i: f"${i['symbol']} {i['price']} {i['change']}",
            lambda i: f"{i['symbol']} {i['price']} · {i['change']}",
        ]

        formatter = random.choice(formats)
        return "\n".join(formatter(item) for item in items)

    @staticmethod
    def vary_emojis(base_emojis: list[str]) -> list[str]:
        """
        Return varied emoji selection.

        Args:
            base_emojis: List of possible emojis

        Returns:
            Random subset of emojis (0-2 emojis)
        """
        count = random.randint(0, min(2, len(base_emojis)))
        return random.sample(base_emojis, count) if count > 0 else []


def monitor_403_errors(func):
    """
    Decorator to monitor 403 Forbidden errors (early muzzling warning).

    Usage:
        @monitor_403_errors
        async def post_tweet(text):
            return await api.create_tweet(text)
    """
    async def wrapper(*args, **kwargs):
        try:
            return await func(*args, **kwargs)
        except Exception as e:
            # Check for 403 Forbidden
            if "403" in str(e) or "Forbidden" in str(e):
                logger.error(f"403 Forbidden detected: {e}")

                # Check if potentially muzzled
                if "muzzled" in str(e).lower() or "write" in str(e).lower():
                    logger.critical(
                        "MUZZLING DETECTED! Check X developer portal immediately."
                    )

            raise

    return wrapper


# Example usage
if __name__ == "__main__":
    import asyncio

    async def example():
        # Initialize
        anti_muzzle = AntiMuzzle(
            max_replies_human=5,
            max_replies_bot=1,
            vip_accounts={"123456789"},  # VIP user IDs
        )

        user_id = "987654321"
        is_bot = False

        # Check if we can reply
        if anti_muzzle.can_reply(user_id, is_bot):
            print(f"✅ Can reply to {user_id}")

            # Add human-like delay
            print("Adding human-like delay...")
            await anti_muzzle.add_human_delay()

            # Simulate posting
            print("Posting tweet...")
            # await api.post_tweet(text)

            # Record the reply
            anti_muzzle.record_reply(user_id)
            print("Reply recorded")
        else:
            print(f"❌ Rate limited for {user_id}")

        # Content variation example
        items = [
            {"symbol": "BTC", "price": "45000", "change": "+5%"},
            {"symbol": "ETH", "price": "2500", "change": "+3%"},
        ]

        varied_content = ContentVariation.vary_list_format(items)
        print(f"\nVaried content:\n{varied_content}")

        # Mention rate limiter example
        mention_limiter = MentionRateLimiter(
            cooldown_seconds=3600,
            bot_handle="@mybot",
        )

        # Cache display names (from X API)
        mention_limiter.cache_display_name("bankrbot", "Bankr")
        mention_limiter.cache_display_name("virtuals_io", "Virtuals Protocol")

        # First mention — allowed
        text1 = mention_limiter.process("Check out @bankrbot for trading!")
        print(f"\nFirst mention: {text1}")
        # -> "Check out @bankrbot for trading!"

        # Second mention within cooldown — replaced with display name
        text2 = mention_limiter.process("Also try @bankrbot for swaps!")
        print(f"Second mention: {text2}")
        # -> "Also try Bankr for swaps!"

    asyncio.run(example())
